KSQL-1361: Add upgrade topology tests

bbejeck opened the pull request:
### Description 
Added testing of an expected topology vs. the generated topology from a KSQL query.

The majority of the changes are in 3 files, `QueryTranslationTest.java`, `EndToEndEngineTestUtil.java`, and the newly added `TopologyFileGenerator.java`

All of the files (88) in the `5_0_expected_topology` directory are just generated topology files for testing.  

Some other points to note:

1. The code only generates a topology for tests that need a topology, i.e tests with expected exceptions don't have an associated topology file.
2. For the tests with AVRO and JSON formatting, there is only one topology file as the data format does not change the topology.


### Testing done 
For testing I did the following:
1. Generated the expected topology files using `TopologyFileGenerator.java`
2. Ran all `ksql-engine` unit tests, including the new check for comparing the expected vs. generated topology
### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

bbejeck commented:
There are some manual steps in using this class but this should be ok as
 we only need to create new topology files at the end of a release cycle.

 The step need to generate topology files:

1. Run this class with the name of the directory.  This requires a relative path from the root
       of the KSQL project i.e "ksql-engine/src/test/resources/DIR_NAME".  The `QueryTranslationTest` 
        expects to find the expected topology files in a directory under src/test/resources.
2. Update the CURRENT_TOPOLOGY_CHECKS_DIR variable in the `QueryTranslationTest` class
     so all tests run against the newly written topology files by default.

guozhangwang commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -520,20 +537,115 @@ private void handleException(final RuntimeException e) {
+          q -> queries.addAll(
+              ksqlEngine.buildMultipleQueries(q, ksqlConfig, query.properties()))
+      );
+     return queries.get(queries.size() - 1).getTopology();
```
Why we assume that only the last statement's topology is needed here?

guozhangwang commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -520,20 +537,115 @@ private void handleException(final RuntimeException e) {
+      final KsqlConfig ksqlConfig = new KsqlConfig(config);
+      final KsqlEngine ksqlEngine = getKsqlEngine(new MockSchemaRegistryClient(), ksqlConfig);
+      final Topology topology = getStreamsTopology(query, ksqlEngine, ksqlConfig);
+      writeExpectedTopologyFile(query.name, topology, topologyDir);
```
Since we use the query.name as the file name, we may need to make sure that 1) these query names would not be changed in the test classes, 2) when tests / queries are removed, their corresponding topology files should also be removed from the corresponding directory.

bbejeck replied:
Changing the query name will cause a test failure and require a rebuild of the expected topologies for that version.  I'll update with some comments to that effect.

However, dropping a test or a query should not cause any issue as that test or query just won't get retrieved from the map, but it's worth test.

guozhangwang commented on ksql-engine/src/test/java/io/confluent/ksql/TopologyFileGenerator.java:
```diff
@@ -0,0 +1,50 @@
+            System.exit(1);
+        }
+
+        EndToEndEngineTestUtil.writeExpectedTopologyFiles(dirPath.toString());
```
Right now we are requiring manual setting the corresponding version number in the parameter when running this cmd, which could be risky. Could we consider letting the version number be retrieved directly from the branch name (e.g. is it possible to read from the pom.xml file)?

bbejeck replied:
ack, will try to do so.

bbejeck commented:
updated this with writing out expected topology directory by pulling the version directly from the `ksql-engine/pom.xml` and improved javadoc

rodesai commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -551,31 +663,47 @@ private static TopologyTestDriver buildStreamsTopology(final Query query,
+    final Map<String, Object> config = getConfigs(new HashMap<>());
     final Properties streamsProperties = new Properties();
     streamsProperties.putAll(config);
     final KsqlConfig ksqlConfig = new KsqlConfig(config);
```
KSQL saves configs alongside queries on the command topic, and uses them to produce the final config used to build the query. That way, if we make a breaking change in the ksql engine, we have a way to preserve the old behavior for stuff that's running. We should replicate this in the test by persisting the config map (as returned by ```KsqlConfig.getKsqlConfigPropsWithSecretsObfuscated```), along with each topology.

Then, before building we should run:

```
Map<String, Object> persistedConfig = <config saved w/ topology>
KsqlConfig ksqlConfig = new KsqlConfig(config).overrideBreakingConfigsWithOriginalValues(persistedConfig)
```

Once we start releasing more often, we'll probably also want to build each query with the current defaults (no overriding) and check that the topology matches whatever the last topology dumped was (with its configs overridden as described above), and then only write a new set of query files out if something has changed.

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/TopologyFileGenerator.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public class TopologyFileGenerator {
+
+    private static final String BASE_DIRECTORY = "ksql-engine/src/test/resources/";
```
at the moment all the resources for a version are held under something like `test/resource/5_0_expected_topology`. It might be nice to store all the expected topologies until a common sub dir, e.g. `test/resources/exected_topology/5.0`. Just so the root `resources` folder doesn't get inundated with versions, and would make it easier in the future if we wanted to iterate over each version.

bbejeck replied:
ack

rodesai commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -551,31 +663,47 @@ private static TopologyTestDriver buildStreamsTopology(final Query query,
-      final TopologyTestDriver testDriver
-          = buildStreamsTopology(query, ksqlEngine, ksqlConfig, streamsProperties);
+      final TopologyTestDriver testDriver = buildStreamsTopologyTestDriver(query, ksqlEngine, ksqlConfig, streamsProperties);
+      assertEquals(query.expectedTopology, query.generatedTopology);
```
Comparing the strings here means that if TopologyDescription.toString changes, then this test breaks. Its also hard to fix the test if it breaks - we'd have to parse the topology description files, and they don't seem very easy to parse. Do you think it makes sense to convert the TopologyDescription to a json-serializable representation, and then write, read, and compare that?

bbejeck replied:
@rodesai I looked into converting the `TopologyDescription` into a JSON format, and it's going to take some work.  Currently using the toString representation is what we use on streams as well as any current streams users. 

What I think could be a better solution 
is to use https://zz85.github.io/kafka-streams-viz/ which is what I used when working on the topology optimization task and made finding issues in a topology

rodesai replied:
That would still be pretty painful right? If we make a change in KafkaStreams to change the string, for example changing indentation, or printing number of partitions for topics, then we'd have to go and update every single test file. I'm fine with using the strings for now, but I still think we should switch to a more readily parsable format at some point.

bbejeck replied:
Yes, I don't disagree with you on this, I think this may take some effort and didn't want to include in this PR.  I've created a separate ticket.

guozhangwang commented:
@bbejeck @rodesai @big-andy-coates I think we may need to discuss a bit what test coverage we want for KSQL-1361. Here's what I have in mind:

There are a few entities we want to check for compatibility:

1. topology (this PR tries to address)
2. configs (as @rodesai mentioned, maybe worth doing in another PR to keep this one small?)

And the scenarios we want to check are that:

a. the above entities stay the same when we push new releases for the same major version.
b. the above entities stay the same when we push a new major version with the optimization config set correctly (this is a temporary approach until we expose APIs for setting internal names).

So I'm think for 1.a) and 1.b), the testing approach would be:

1. Whenever we release a new major version (let's say 5.1), the release process will automatically call topology generator to write a new set of topologies under a new folder (e.g. `5.1`). I.e. the topology generator will only be executed once per release, followed by a dummy commit to check in the topology files.

As a result, only branch `5.1.x`, `5.1.0-post` and `master` will contain topology folder `5.1`; note `master` will contain all folders up to the latest release version.

2. We will run the compatibility test as part of our testbrake, on branch `5.1.x` to compare the generated topology with the stored topology in `5.1`.

3. We will also run the compatibility test on branch `5.1.x`, with the optimization config set to `none` to compare the generated topology with the stored topology in `5.0`.

@rodesai you mentioned that `... then only write a new set of query files out if something has changed.` I thought we would not expect this to happen and if it happened should be a test failure. Could you elaborate a bit on this?

4. If we add more tests in the future releases (again, say in `5.1.x` branch), we should create a PR against `5.1.0-post` instead, that includes a) the added query, b) query's generated topology with the code `5.1.0-post`. And then merge-down to `5.1.x`. Note that unless we intend to do so, the corresponding query would not exist in 5.0.x branch, so the test 3) should handle the case when older folders does not have the corresponding topology file, indicating this query does not exit in that version.


For 2, we can consider following the same workflow, but would like to hear your opinions.

rodesai commented:
> There are a few entities we want to check for compatibility:
topology (this PR tries to address)
configs (as @rodesai mentioned, maybe worth doing in another PR to keep this one small?)

My concern is less about checking that the configs are the same, but more about making sure we test how KSQL handles compatibility (not just Streams).

Let's use an example. Suppose I have the following stream and table:

```sql
CREATE STREAM S (SK STRING, SV STRING) ...;
CREATE TABLE T (TK STRING , TV1 STRING, TV2 STRING, ... TV10 STRING)
```

And I do a join:
```
CREATE STREAM J AS SELECT S.*, T.V1 FROM S JOIN T ON S.SK = T.TK
```

Currently, KSQL will build a state store that materializes all the fields in T. We could optimize by projecting T.TK and T.V1 first. However, if we did that for queries started before the optimization was implemented we might break them.

The current solution for such a situation in KSQL is to toggle the optimization via a config that gets _defaulted_ to ```off``` for statements read from the command topic, and set to ```on``` for statements written to it. So if the statement was written before the optimization was implemented, when we read it back the setting will be ```off```. But for new statements, its explicitly set to ```on```. My point was that we should mimic this process by writing out the current config spec for each query when we write out the topology, and read it back with all compatibility-breaking configs _defaulted_ to off before building the topology at test time.

> @rodesai you mentioned that ... then only write a new set of query files out if something has changed. I thought we would not expect this to happen and if it happened should be a test failure. Could you elaborate a bit on this?

This was more of an in-the-future thing. But if at some point we wanted to be able to release much more frequently (e.g. every day or week), we probably don't want to write out a new set of topologies every day or week. We just want to write them out when we make a compatibility-breaking change.

bbejeck commented on ksql-engine/src/test/java/io/confluent/ksql/QueryTranslationTest.java:
```diff
@@ -26,27 +34,34 @@
-import org.junit.runner.RunWith;
-import org.junit.runners.Parameterized;
+
+import static io.confluent.ksql.EndToEndEngineTestUtil.ExpectedException;
```
I had installed GoogleStyle previously, but I've updated it again from the link above.  
FWIW I think I might have had "sort imports" checked on my IntelliJ commit dialog.  I've updated the GoogleStyle code xml file and made sure un-select the sort imports dialog!

bbejeck commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -520,20 +537,115 @@ private void handleException(final RuntimeException e) {
-    return new TopologyTestDriver(queries.get(queries.size() - 1).getTopology(),
+
+  static void writeExpectedTopologyFiles(final String topologyDir) throws IOException {
+    final List<Query> queryList = QueryTranslationTest.buildQueryList().stream()
```
Ack. Makes a lot of sense, I don't know why I didn't do this in the first place!

bbejeck commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -520,20 +537,115 @@ private void handleException(final RuntimeException e) {
+    final Random randomPort = new Random();
+    queryList.forEach(query -> {
+      final Map<String, Object> config = getConfigs(null);
+      config.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:" + randomPort.nextInt(4000));
```
I needed a unique port for each invocation during the file generation. I tried several options, but the process was getting stuck after writing the first file.

bbejeck commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -520,20 +537,115 @@ private void handleException(final RuntimeException e) {
+        try {
+            final String updatedQueryName = formatQueryName(queryName);
+
+            if (!Paths.get(topologyDir, updatedQueryName).toFile().exists()) {
```
ack

bbejeck commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -520,20 +537,115 @@ private void handleException(final RuntimeException e) {
+        }
+    }
+
+  static String formatQueryName(String originalQueryName) {
```
ack

bbejeck commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -520,20 +537,115 @@ private void handleException(final RuntimeException e) {
+
+  private static String readTopologyFile(String dir) {
+    final StringBuilder topologyFileBuilder = new StringBuilder();
+    try(final BufferedReader reader =
```
ack

bbejeck commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -520,20 +537,115 @@ private void handleException(final RuntimeException e) {
+          q -> queries.addAll(
+              ksqlEngine.buildMultipleQueries(q, ksqlConfig, query.properties()))
+      );
+     return queries.get(queries.size() - 1).getTopology();
```
ack

bbejeck commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -520,20 +537,115 @@ private void handleException(final RuntimeException e) {
+      return expectedTopologies;
+  }
+
+  private static String readTopologyFile(String dir) {
```
ack

guozhangwang commented:
The change LGTM. Leaving to @rodesai to make a final pass.

bbejeck commented:
rebased this

rodesai commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -520,20 +550,140 @@ private void handleException(final RuntimeException e) {
+                                                                   final KsqlConfig ksqlConfig,
+                                                                   final Properties streamsProperties) {
+
+    final KsqlConfig maybeUpdatedConfigs = query.persistedProperties ()== null ? ksqlConfig :
```
nit: spacing

rodesai commented on ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java:
```diff
@@ -435,6 +449,22 @@ public String getName() {
+       this.persistedProperties = persistedProperties;
+    }
+
+    public Map<String, String> persistedProperties() {
```
nit: consider returning Optional<Map<String, String>> here.

bbejeck commented:
updated per comments, added better descriptions for running file generation and running topology validation (QueryTranslationTest) against older released versions of KSQL
